docs/architecture.md

```diff
@@ -0,0 +1,176 @@
+
+## Requestor Implementation
+
+The requestor implementation consists of the RequestManager which makes and tracks requests and the AsyncLoader subsystem which manages incoming responses. These systems work in concert to verify responses by performing a local selector traversal that is backed by network response data.
+
+When a request is initiated, the RequestManager sends the request across the network, begins tracking it, initiates a query to verify the responses from the network, and initiates a thread to ultimately collect and return verified data to the caller. Most of these processes are straightforward, but verifying the response is not simple. We need to execute an IPLD selector query backed by network data, which will arrive asynchronously and potentially out of order.
+
+To accomplish this, we delegate to the AsyncLoader, which is responsible for ingesting responses from the network, and providing them to the local IPLD selector traversal operation as it loads links.
+
+The AsyncLoader needs to take in blocks and response metadata from the network, put that data in an in memory cache until either:
+
+- The local traversal actually requests that data, which validates the block, and therefore tells the async loader it can move to permanent storage.
+- The local traversal completes, at which point, any blocks that haven't been verified can be dropped, cause they've been shown not to be the correct responses.
+
+The AsyncLoader has to also manage requests for link-loads coming in form the local traversal, which may come in before OR after the relevant blocks are received from the network. Moreover, if a block is already in local storage for whatever reason, there's no need to hold up the local traversal on a network response.
+
+The following process outlines the basic process for loading links asynchronously from the network:
+
+![Async Loading Process](async-loading.png)
+
+The main components that make up the AsyncLoader are:
+
+* The UnverifiedBlockStore -- basically just a temporary cache to hold blocks, and write them to permaneant storage as they are verified
+* The ResponseCache -- essentially just the UnverifiedBlockStore + some link tracking, so that the BlockStore is properly pruned as requests expire
+* The LoadAttemptQueue -- Basically a queue to track attempts to load links. A link load can have one of three results --
+- It can load successfully,
+- It can load and knowingly fail (i.e responder tells you it is for sure missing a link)
+- It can be indeterminate -- you don't yet have the block in the response cache or in local storage, but the responder is still sending responses and hasn't indicated it doesn't have the block
```

Comment:
```suggestion
  - It can be indeterminate -- you don't yet have the block in the response cache or in local storage, but the responder is still sending responses and hasn't indicated it doesn't have the block
```